fix: SceneEventProgress OnComplete not invoked on timeouts [MTT-4772]

NoelStephensUnity opened the pull request:
This PR addresses several issues with `SceneEventProgress` that could result in it never completing as well as several scenarios where it would never time out.

[MTT-4772](https://jira.unity3d.com/browse/MTT-4772)

## Changelog
- Fixed: issue where `SceneEventProgress` would not complete if a client late joins while it is still in progress.
- Fixed: issue where `SceneEventProgress` would not complete if a client disconnects.
- Fixed: issues with detecting if a `SceneEventProgress` has timed out. 

## Testing and Documentation
- Includes `SceneEventProgress` specific integration tests.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

MTT-4772
This is still WIP.
Either there is an issue with the integration test and disconnecting a client while loading or a bug with some of the updates to handling disconnecting clients during a scene event.

NoelStephensUnity committed:
fix

Need to track any clients that disconnected to add them to the list of clients that did not complete the scene event.

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
fix and style

Reducing the logic used to determine when the SceneEventProgress is finished.
Renamed a few properties and methods for code readability purposes.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/ISceneManagerHandler.cs:
```diff
@@ -11,7 +11,7 @@ namespace Unity.Netcode
     {
         // Generic action to call when a scene is finished loading/unloading
-        struct SceneEventAction
+        class SceneEventAction
```
Just in case you were wondering about this change, making it a class simplifies how actions are tracked during integration tests and avoids some issues when they are in the queue to be processed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
fix and style

Added additional comments for clarity
Fixed the logic behind determining when the SceneEventProgress should complete.

NoelStephensUnity committed:
test

Added several tests to verify that SceneEventProgress will complete if all clients finished the scene event, if clients disconnect while it is still in progress, will time out if one or more clients never finish, and finally will still complete if one or more clients late join while it is still in progress.

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
update

MTT-4772
Updating changelog with what was fixed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
test

Realized using random to determine if a client should or should not do something could result in no clients having done something (i.e. disconnect, late join, never finish).
This update assures two clients will perform the action being tested.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/ISceneManagerHandler.cs:
```diff
@@ -11,7 +11,7 @@ namespace Unity.Netcode
     {
         // Generic action to call when a scene is finished loading/unloading
-        struct SceneEventAction
+        class SceneEventAction
```
Just curious why this change was necessary. Not sure how often these are created so I don't know what the GC cost is, but changing to a class certainly adds more than 0 GC, right?

NoelStephensUnity replied:
Yeah.. that should have been a class.
The previous scene event progress was returning a false positive but the completed was never triggering.

NoelStephensUnity replied:
@ShadauxCat 
SceneEventAction no longer exists.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventProgress.cs:
```diff
@@ -171,37 +221,40 @@ internal void SetSceneLoadOperation(AsyncOperation sceneLoadOperation)
         internal void SetSceneLoadOperation(ISceneManagerHandler.SceneEventAction sceneEventAction)
         {
-            sceneEventAction.Completed = SetComplete;
+            sceneEventAction.Completed = SetLoadCompleted;
```
Ah, is this why SceneEventAction was changed to a class? Because the struct passed here was a copy and this was basically ending up NOP?

Would it be possible to take the struct by `ref` here and avoid making it a class?

NoelStephensUnity replied:
Good point on the ref...let me check that... and yes that was why I semi-punted since it is really only used during integration tests.

NoelStephensUnity replied:
After looking at this, the cost of the class malloc is tiny and occurs only during a scene load or unload (when there is typically much larger allocations).  
I could try to refactor this whole process but I am not sure the savings we will get during loading and unloading would be worth it.

NoelStephensUnity replied:
Did a refactor of that entire crazy process.
You get a 💯 for calling that out... forced me to further think about "what else is wrong with this picture".
👍

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
refactor

Completely removed the entire SceneEventAction crazy.
Replaced it with an even more simplified version of SceneEventProgress.
We will no longer allocate memory for SceneEventAction.  :laughing:
(and it is cleaner)  :+1:

NoelStephensUnity committed:
style

fixing grammar in a comment
removing two CR/LFs

NoelStephensUnity committed:
style

moving a comment above a few lines of code (otherwise it doesn't make sense).

NoelStephensUnity committed:
style

>.<
Last comment adjustment

NoelStephensUnity committed:
update

Renaming the SetSceneAsyncOperation to just SetAsyncOperation.

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772

NoelStephensUnity committed:
Merge branch 'develop' into fix/sceneeventprogress-timedout-and-clients-connected-logic-mtt-4772
